ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionLogUpdatesConversions.scala

```diff
@@ -182,43 +194,40 @@ private[events] object TransactionLogUpdatesConversions {
             eventId = exercisedEvent.eventId.toLedgerString,
             contractId = exercisedEvent.contractId.coid,
             templateId = Some(LfEngineToApi.toApiIdentifier(exercisedEvent.templateId)),
-            witnessParties = requestingParties.view.filter(exercisedEvent.flatEventWitnesses).toSeq,
+            witnessParties =
+              requestingParties.iterator.filter(exercisedEvent.flatEventWitnesses).toSeq,
           )
         )
       )
   }
 
   object ToTransactionTree {
-    def apply(
-        tx: TransactionLogUpdate.Transaction,
-        requestingParties: Set[String],
+    def filter(requestingParties: Set[Party]): Filter = transaction => {
+      val filteredForVisibility =
+        transaction.events
+          .filter(transactionTreePredicate(requestingParties))
+
+      Option.when(filteredForVisibility.nonEmpty)(transaction.copy(events = filteredForVisibility))
+    }
+
+    def toApiTransaction(
+        requestingParties: Set[Party],
         verbose: Boolean,
         lfValueTranslation: LfValueTranslation,
     )(implicit
         loggingContext: LoggingContext,
         executionContext: ExecutionContext,
-    ): Future[Option[TransactionTree]] = {
-      val filteredForVisibility = tx.events
-        .filter(TransactionTreePredicate(requestingParties))
-
-      if (filteredForVisibility.isEmpty) Future.successful(None)
-      else
+    ): ToApi[GetTransactionTreesResponse] =
+      transaction =>
         Future
-          .traverse(filteredForVisibility)(
+          .traverse(transaction.events)(
```

Comment:
Same here as I commented on the same function for flat above in this file (wrapping the traverse in a Future)

Reply:
Thanks, good remark. Will update in the subsequent performance improvement PR.